### ISSUE TITLE
Redesign index-storage interactions

### ISSUE THREAD

**Issue by kirillt** [ARK-Builders/arklib-android#42]
- [x] Index-related classes should not be coupled with storages too tightly.
- [x] Index-related classes should not call any factories or storages:
       1. Storages should consume updates directly or via presenters.
       2. Factories should be invoked by storages when they don't have data or it is outdated.

Examples of improper classes interaction:

`ResourceIndexRepo` depends on `MetadataStorage`
```
class ResourceIndexRepo(
    private val foldersRepo: FoldersRepo,
    //todo any storage should be attached next to index, not into it
    private val metadataStorageRepo: MetadataStorageRepo,
    private val messageFlow: MutableSharedFlow<Message>,
...
```

`PlainIndex` during index updates processing invokes `GeneralMetadataFactory`:
```
    private suspend fun handleUpdate(update: UpdatedResourcesId) {
        update.deleted.forEach { ... }
        val added = ...
        ...

        //todo MetadataStorage must manage Metadata for resources
        // the same as TagStorage manages their tags
                GeneralMetadataFactory.compute(path, resource)
                    .onFailure {
                        Log.e(
                                RESOURCES_INDEX,
                                "Could not detect kind for " +
                                        path.absolutePathString()
                            )
                        messageFlow.emit(Message.KindDetectFailed(path))
                    }
                    .map { metadata ->
                        resource.metadata = metadata
                        resource
                    }
...
```

`ResourceIndexRepo` during initial index loading invokes `GeneralMetadataFactory`:
```
    internal suspend fun providePlainIndex(
        root: Path
    ): PlainIndex = provideMutex.withLock {
        ...
        //todo MetadataStorage must manage Metadata for resources
        // the same as TagStorage manages all tags
                GeneralMetadataFactory.compute(path, resource)
                    .map { metadata ->
                        resource.metadata = metadata
                        resource
                    }
```

**Comment by kirillt** [comment]
This also causes some problems:
```
data class Resource(
    val id: ResourceId,
    val name: String,
    val extension: String,
    val modified: FileTime,
    var metadata: Metadata?, //todo: get rid of it
) {
```

It's better to completely decouple `Resource` from `Metadata`.
`Metadata` should be retrieved from `MetadataStorage`, same as `Tags` are retrieved from `TagsStorage`.

**Comment by kirillt** [comment]
Solved in #52
